File: Easy Problems/Ransom_Note.py
'''
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
Each letter in magazine can only be used once in ransomNote.

Example 1:
Input: ransomNote = "a", magazine = "b"
Output: false

Example 2:
Input: ransomNote = "aa", magazine = "ab"
Output: false

Example 3:
Input: ransomNote = "aa", magazine = "aab"
Output: true

Constraints:
1 <= ransomNote.length, magazine.length <= 105
ransomNote and magazine consist of lowercase English letters.
'''
from collections import Counter


class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        ransom_counter = Counter(ransomNote)
        magazine_counter = Counter(magazine)

        for char, count in ransom_counter.items():
            if magazine_counter[char] < count:
                return False
        return True


ransomNote = "aab"
magazine = "baa"
solution = Solution()
print(solution.canConstruct(ransomNote, magazine))

File: Easy Problems/Ransom_Note.js
/*
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
Each letter in magazine can only be used once in ransomNote.

Example 1:
Input: ransomNote = "a", magazine = "b"
Output: false

Example 2:
Input: ransomNote = "aa", magazine = "ab"
Output: false

Example 3:
Input: ransomNote = "aa", magazine = "aab"
Output: true

Constraints:
1 <= ransomNote.length, magazine.length <= 105
ransomNote and magazine consist of lowercase English letters.
*/

/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
    let ransomCount = {};
    let magazineCount = {};

    for (let i of ransomNote) {
        if (ransomCount[i]) {
            ransomCount[i] += 1;
        } else {
            ransomCount[i] = 1;
        }
    }

    for (let i of magazine) {
        if (magazineCount[i]) {
            magazineCount[i] += 1;
        } else {
            magazineCount[i] = 1;
        }
    }

    for (let [char, count] of Object.entries(ransomCount)) {
        if (!magazineCount[char] || magazineCount[char] < count) {
            return false;
        }
    }

    return true;
};

let ransomNote = "aa";
let magazine = "aab";
console.log(canConstruct(ransomNote, magazine));